java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -139,3 +140,32 @@ private class StringBuilderTaintPreservingCallable extends TaintPreservingCallab
     sink = -1
   }
 }
+
+/** Method access to external inputs of `android.os.BaseBundle` object. */
+class GetBundleExtraMethodAccess extends MethodAccess {
+  GetBundleExtraMethodAccess() {
+    this.getMethod().getName().regexpMatch("get\\w+") and
+    this
+        .getMethod()
+        .getDeclaringType()
+        .getASupertype*()
+        .hasQualifiedName("android.os", "BaseBundle")
+  }
+}
+
+/**
+ * Holds if `n1` to `n2` is a dataflow step between the extra getter method and its caller `Bundle`.
+ */
+private predicate bundleExtraStep(DataFlow::ExprNode n1, DataFlow::ExprNode n2) {
+  exists(GetBundleExtraMethodAccess ma |
+    n1.asExpr() = ma.getQualifier() and
+    n2.asExpr() = ma
+  )
+}
+
+/** Additional taint step to consider when taint tracking Android intent extra related data flows. */
+class AndroidExtraSourceAdditionalTaintStep extends AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    bundleExtraStep(node1, node2)
+  }
+}
```

Comment:
Re: TaintPreservingCallable, I found all of this could be replaced with

```codeql
class BundleGetterMethods extends Method, TaintPreservingCallable {
  BundleGetterMethods() {
    getDeclaringType().hasQualifiedName("android.os", ["BaseBundle", "Bundle"]) and
    getName().matches("get%")
  }

  override predicate returnsTaintFrom(int arg) { arg = -1 }
}
```
Note that `getExtras()` is already a TaintPreservingCallable. We should also store steps like this in `Android.qll`, not the top-level `FlowSteps.qll`.

Reply:
Thanks @smowton for the code. I've updated the query and added the following test cases:
- Changed the query of CWE-749 to use the new <code>RemoteFlowSource</code>
- Added an AndroidManifest.xml file with six new activities of different configuration (explicit <code>android:exported="true|false"</code>, with/without intent filter, and implicit hide/export
- Added a <code>BroadcastReceiver</code> to test <code>AndroidReceiveIntentMethod</code>

As you expected, utility method that takes an Activity a then calls a.getIntent().getStringExtra("xyz"), which is passed an exported component, should be flagged but is not because of how we phrased the source. This is demonstrated in <code>java/ql/test/experimental/query-tests/security/CWE-749/UnsafeActivity4.java</code>, which invokes the utility method <code>IntentUtils.getIntentUrl(activity)</code>. The same is to <code>a.getIntent().getExtras().getString("xyz")</code> How could we tackle this scenario? Thanks.